src/processor/LakeOfFHIR.ts

```diff
@@ -121,4 +121,36 @@ export class LakeOfFHIR implements utils.Fishable {
       );
     }
   }
+
+  /**
+   * All definitions should have a resourceType and id. If any definition is missing an id,
+   * we add one either based on the name or with a simple counter.
+   * Logs a warning if it finds any definitions without an id
+   */
+  assignMissingIds() {
+    const createdIdPaths: string[] = [];
+    let generatedId = 0;
+    this.docs.forEach((d, index) => {
+      if (d.content.id == null) {
+        // Try to be smart and set the id to the existing name
+        d.content.id = d.content.name;
+        if (d.content.id == null) {
+          // If there is no existing name, generate an id with a counter
+          d.content.id = `id-${generatedId++}`;
```

Comment:
I've found that if there is no existing name, then we don't actually generate a profile at all. Are you seeing the same thing? Are there cases where we generate entities when the `name` is not present? 

This may not necessarily be in the scope of this PR, but if we aren't handling entities that don't have names, then I'm not sure we need the code to handle them here. And confusingly, we log a message indicating that GoFSH created an ID for a profile that then does not exist.

Reply:
Hmm that is a great point and a great question. I hadn't actually noticed that when implementing this, but now that I'm looking at it, I notice that we are generating CodeSystems, ValueSets, and Instances that don't have a `name`, but not StructureDefinitions or ImplementationGuides. I'm not entirely sure what the right way forward is, besides that it should probably be consistent. Now that we're setting `id`s on everything, it feels like everything should be generated now, rather than changing some so they do not generate. But I'm also interested in what you and @mint-thompson think. (And anyone else on the team too.)

The other thing that feels somewhat related is that when we do the roundtrip `-f`, the names of the files change when SUSHI builds them because SUSHI uses the `resourceType-id` scheme, which, when you're missing an id on the original input, it is not included in the name either. I'm not sure how to fix that either though. And I'm not sure if either of this should be a separate task or if it makes sense to just do them now.

Reply:
> I notice that we are generating CodeSystems, ValueSets, and Instances that don't have a name, but not StructureDefinitions or ImplementationGuides.

This is only a problem if people give us bad FHIR.  Whereas `name` is _optional_ in `ValueSet`, `CodeSystem`, and many `Instance`s, it is _required_ (`1..1`) on `StructureDefinition` and `ImplementationGuide`.  So the only way we'd ever end up having one without a name is if we started with invalid FHIR definitions.  (Which has happened before, so...).

> Now that we're setting ids on everything, it feels like everything should be generated now, rather than changing some so they do not generate.

If it's a simple change, that would be fine w/ me.  As I said above, lacking `name` on `StructureDefinition` or `ImplementationGuide` is technically invalid -- so if someone gives us invalid FHIR, I'm OK with making up a name so it is valid and we can write it.

> The other thing that feels somewhat related is that when we do the roundtrip -f, the names of the files change when SUSHI builds them because SUSHI uses the resourceType-id scheme...

Yeah, I think I'm OK with this.  FSHing trip should be useful to the extent that it can be so, but I don't think we should change SUSHI's filename scheme in order to satisfy FSHing trip.  If anything, we should first see if there are things we can do in FSHing trip so that it knows how old and new files map to each other and can correctly line them up when doing the diffs.

Reply:
So slight snag. After talking at stand up, I was going to just always add a `name` if we tried setting and `id` and there was no `name` set either. But I don't actually think we can take that simplistic of an approach because not all definitions will have a name because not all `Instances` will have a name. So I think that still leaves us with the discrepancy that Nick pointed out - we might log a warning saying we changed the id of a definition that does not actually get exported as FSH because it was also missing a `name`.

So I am back to not being entirely sure what the right approach is. The two thoughts I have are:
- I could log a warning when we skip over StructureDefinitions without names. Maybe then the first warning about updating the `id` would be less confusing if there is a later warning saying why that definition didn't actually get created.
- I could check if the definition is a type `StructureDefinition` or `ImplementationGuide` before setting an `id` in this function since those seem to be the two that require `name` when processing, and then we wouldn't log a warning for that file.

Reply:
I'm not sure I fully understand the issue.  If you can check if the definition is a `StructureDefinition` or `ImplementationGuide` in order to suppress the warning, then can't you check if the definition is a `StructureDefinition` or `ImplementationGuide` to decide if it is OK to set a name?

I.e., even if you didn't do it in this function, couldn't you create a `LakeOfFHIR.assignMissingNames()` function that is just like this but assigns missing names to SDs and IGs that are missing names?  Or... I think Nick said that this doesn't happen w/ ValueSets/CodeSystems that are missing names.  So... how do they handle it?  Do they handle it in their distinct processors or exporters?

Reply:
Alright! I think I've got it! (although, who really knows). Instead of checking for and adding `name`s as needed in `assignMissingIds()`, I updated the `StructureDefinition.isProcessableStructureDefinition` function to align more closely with the checks in `CodeSystem` and `ValueSet`. An SD is now considered processable if it has a `resourceType` of `'StructureDefinition'` and if either the `name` or `id` is defined. Then, it uses the input's `name` or `id` to set the Exportable's name. Technically we shouldn't need to worry about having a missing `name` since it is required, but hopefully this will appropriately catch the cases where we have invalid FHIR.

So now I _think_ this thread is set! Let me know what you think!